File: integrations/claude-managed-agents/python/ag_ui_claude_managed_agents/agent.py
"""ManagedAgentsAgent: an AG-UI agent backed by Claude Managed Agents."""

import asyncio
from collections.abc import AsyncIterator, Callable, Sequence
from dataclasses import dataclass, field
from typing import Any, ClassVar, Literal

from ag_ui.core import (
    BaseEvent,
    CustomEvent,
    RunAgentInput,
    RunErrorEvent,
    RunFinishedEvent,
    RunStartedEvent,
    StateSnapshotEvent,
)
from anthropic import AsyncAnthropic

from ._util import maybe_await
from .constants import DEFAULT_TURN_TIMEOUT_S
from .sessions import InMemorySessionStore
from .tools import custom_tool_from, normalize_tool_name
from .turn import Emit, run_turn
from .types import BackendTool, SessionRecord, SessionStore, TurnOutcome

_DONE = object()


def _user_text(message: Any) -> str:
    """The text of a user message (string content or multimodal parts).

    Non-text parts (images, documents) are dropped: a message carrying only
    those has no text and errors as an empty run.
    """
    content = getattr(message, "content", None)
    if isinstance(content, str):
        return content
    parts: list[str] = []
    for part in content or []:
        text = getattr(part, "text", None)
        if isinstance(text, str):
            parts.append(text)
    return "".join(parts)


@dataclass
class _Outbound:
    events: list[dict[str, Any]] = field(default_factory=list)
    still_parked: list[str] = field(default_factory=list)
    last_user_message_id: str | None = None


@dataclass
class _RunState:
    session_id: str | None = None
    busy_key: str | None = None
    """The busy-thread gate this run holds, released when the run unwinds."""


class ManagedAgentsAgent:
    """An AG-UI agent backed by Claude Managed Agents.

    Each AG-UI thread maps to one managed session; each run drives one turn
    of that session.
    """

    # Keyed by session-store identity: the store is the unit of tenancy, so
    # agents sharing a store serialize runs per thread (even across instances),
    # while per-caller stores keep one caller's runs from blocking another's.
    # Keys within a store's set are scoped to the managed agent.
    _busy_threads: ClassVar[dict[int, set[str]]] = {}

    def __init__(
        self,
        *,
        managed_agent_id: str,
        environment_id: str,
        agent_version: int | None = None,
        client: AsyncAnthropic | None = None,
        session_store: SessionStore | None = None,
        backend_tools: list[BackendTool] | None = None,
        session_title: Callable[[str], str] | None = None,
        tool_confirmation: Literal["allow", "deny"] | None = None,
        turn_timeout_s: float = DEFAULT_TURN_TIMEOUT_S,
        stream_deltas: bool = True,
    ) -> None:
        self.managed_agent_id = managed_agent_id
        self.environment_id = environment_id
        self.agent_version = agent_version
        self.client = client if client is not None else AsyncAnthropic()
        self.store: SessionStore = (
            session_store if session_store is not None else InMemorySessionStore()
        )
        # Keyed by normalized name; on a normalized-name collision (e.g. "search
        # web" and "search_web") the last tool wins.
        self.backend_tools: dict[str, BackendTool] = {
            normalize_tool_name(tool.name): tool for tool in backend_tools or []
        }
        self.session_title = session_title
        self.tool_confirmation = tool_confirmation
        self.turn_timeout_s = turn_timeout_s
        self.stream_deltas = stream_deltas
        # Strong references to in-flight worker tasks so they are not collected.
        self._tasks: set[asyncio.Task[None]] = set()

    async def run(self, input: RunAgentInput) -> AsyncIterator[BaseEvent]:  # noqa: A002 - matches AG-UI adapters
        """Run one turn for `input`, yielding AG-UI events."""
        queue: asyncio.Queue[Any] = asyncio.Queue()
        state = _RunState()
        worker = asyncio.create_task(self._drive(input, queue.put_nowait, state))
        self._tasks.add(worker)
        worker.add_done_callback(self._tasks.discard)
        worker.add_done_callback(lambda _task: queue.put_nowait(_DONE))
        try:
            while True:
                item = await queue.get()
                if item is _DONE:
                    return
                yield item
        finally:
            # The client went away (or iteration stopped): cancel the turn,
            # which interrupts the session.
            if not worker.done():
                worker.cancel()

    async def _drive(self, input: RunAgentInput, emit: Emit, state: _RunState) -> None:  # noqa: A002
        try:
            await asyncio.wait_for(
                self._run_turn_for_input(input, emit, state),
                timeout=self.turn_timeout_s,
            )
        except asyncio.TimeoutError:
            await self._interrupt(state.session_id)
            emit(
                RunErrorEvent(
                    message=(
                        f"The turn exceeded the {self.turn_timeout_s:g}s limit"
                        " and was interrupted."
                    ),
                    code="turn_timeout",
                )
            )
        except asyncio.CancelledError:
            # Client disconnected mid-turn: stop the session; there is nobody
            # left to tell. This runs before the busy gate is released (see
            # `finally`), so a user who resends right away is not interrupted
            # by this late stop.
            await self._interrupt(state.session_id)
            raise
        except Exception as err:  # noqa: BLE001 - surfaced to the client as RUN_ERROR
            emit(
                RunErrorEvent(message=str(err) or "The run failed.", code="run_failed")
            )
        finally:
            # Release the per-thread gate only after any interrupt above has
            # been posted, never before.
            if state.busy_key is not None:
                busy = ManagedAgentsAgent._busy_threads.get(id(self.store))
                if busy is not None:
                    busy.discard(state.busy_key)
                    if not busy:
                        del ManagedAgentsAgent._busy_threads[id(self.store)]

    async def _interrupt(self, session_id: str | None) -> None:
        if not session_id:
            return
        try:
            await self.client.beta.sessions.events.send(
                session_id, events=[{"type": "user.interrupt"}]
            )
        except Exception:  # noqa: BLE001 - best-effort interrupt
            pass

    async def _run_turn_for_input(
        self, input: RunAgentInput, emit: Emit, state: _RunState
    ) -> None:  # noqa: A002
        thread_id, run_id = input.thread_id, input.run_id
        busy_key = self._thread_key(thread_id)
        emit(RunStartedEvent(thread_id=thread_id, run_id=run_id))
        if input.state is not None:
            emit(StateSnapshotEvent(snapshot=input.state))

        if busy_key in ManagedAgentsAgent._busy_threads.get(id(self.store), ()):
            emit(
                RunErrorEvent(
                    message="A run is already in progress on this thread.",
                    code="run_in_progress",
                )
            )
            return
        # Check for something sendable before touching the API, so a malformed
        # run does not create an orphan session.
        if not self._has_sendable_content(input.messages):
            emit(
                RunErrorEvent(
                    message="There is nothing to send: this run has no user message or tool result.",
                    code="empty_run",
                )
            )
            return

        # Hold the per-thread gate for the rest of the run; `_drive` releases
        # it after any interrupt has been posted.
        ManagedAgentsAgent._busy_threads.setdefault(id(self.store), set()).add(busy_key)
        state.busy_key = busy_key

        record = await self._get_or_create_session(thread_id, input, emit)
        if record is None:
            emit(
                RunErrorEvent(
                    message="There is nothing to send: a tool result arrived for a thread with no session.",
                    code="tool_result_without_session",
                )
            )
            return
        state.session_id = record.session_id
        await self._sync_client_tools(record, input.tools or [])

        outbound = self._outbound_events(record, input.messages)
        if not outbound.events:
            emit(
                RunErrorEvent(
                    message="There is nothing new to send: no user message or tool result in this run.",
                    code="nothing_to_send",
                )
            )
            return

        store_key = thread_id

        # Some parked tool calls are still unanswered: post what we have and
        # stay parked instead of waiting on a session that will not resume.
        if outbound.still_parked:
            await self.client.beta.sessions.events.send(
                record.session_id, events=outbound.events
            )
            record.pending_client_tool_use_ids = outbound.still_parked
            record.last_user_message_id = (
                outbound.last_user_message_id or record.last_user_message_id
            )
            await maybe_await(self.store.set(store_key, record))
            emit(RunFinishedEvent(thread_id=thread_id, run_id=run_id))
            return

        # Persist each delivery as soon as it lands, so a failure or
        # interruption later in the turn does not re-post it next run: the
        # tool results resume the session even if the follow-ups then fail.
        async def on_results_sent() -> None:
            record.pending_client_tool_use_ids = []
            await maybe_await(self.store.set(store_key, record))

        async def on_follow_ups_sent() -> None:
            if outbound.last_user_message_id:
                record.last_user_message_id = outbound.last_user_message_id
            await maybe_await(self.store.set(store_key, record))

        outcome = await run_turn(
            client=self.client,
            session_id=record.session_id,
            outbound=outbound.events,
            on_results_sent=on_results_sent,
            on_follow_ups_sent=on_follow_ups_sent,
            client_tools=self._client_tools(input.tools or []),
            backend_tools=self.backend_tools,
            tool_confirmation=self.tool_confirmation,
            stream_deltas=self.stream_deltas,
            emit=emit,
        )

        await self._record_outcome(thread_id, record, outcome)
        if outcome.status != "errored":
            emit(RunFinishedEvent(thread_id=thread_id, run_id=run_id))

    def _client_tools(self, client_tools: Sequence[Any]) -> dict[str, str]:
        """Normalized frontend tool name -> its original AG-UI name.

        On a normalized-name collision the last tool wins.
        """
        return {normalize_tool_name(tool.name): tool.name for tool in client_tools}

    def _has_sendable_content(self, messages: Sequence[Any]) -> bool:
        """Whether the run carries a user message with text or a tool result."""
        for message in messages:
            role = getattr(message, "role", None)
            if role == "tool":
                return True
            if role == "user" and _user_text(message).strip():
                return True
        return False

    def _thread_key(self, thread_id: str) -> str:
        """Shared-state key scoped to this managed agent so distinct agents never collide."""
        return f"{self.managed_agent_id}:{thread_id}"

    async def _record_outcome(
        self, thread_id: str, record: SessionRecord, outcome: TurnOutcome
    ) -> None:
        if outcome.status == "errored" and outcome.session_ended:
            await maybe_await(self.store.delete(thread_id))
            return
        if outcome.status != "parked":
            return
        record.pending_client_tool_use_ids = list(outcome.client_tool_use_ids)
        await maybe_await(self.store.set(thread_id, record))

    def _outbound_events(
        self, record: SessionRecord, messages: Sequence[Any]
    ) -> _Outbound:
        """Work out what to post into the session for this run: results for any
        tool calls the frontend was asked to run, plus every user message not
        yet delivered (in order).
        """
        events: list[dict[str, Any]] = []
        pending = list(record.pending_client_tool_use_ids)

        for message in messages:
            if getattr(message, "role", None) != "tool":
                continue
            tool_call_id = getattr(message, "tool_call_id", None)
            if tool_call_id not in pending:
                continue
            error_text = getattr(message, "error", None)
            result_text = "\n".join(
                part for part in (message.content or "", error_text or "") if part
            )
            events.append(
                {
                    "type": "user.custom_tool_result",
                    "custom_tool_use_id": tool_call_id,
                    "content": [{"type": "text", "text": result_text}],
                    "is_error": bool(error_text),
                }
            )
            pending.remove(tool_call_id)

        # User messages after the last delivered one; on first contact, just the newest.
        last_user_message_id: str | None = None
        user_messages = [
            message for message in messages if getattr(message, "role", None) == "user"
        ]
        delivered_index = next(
            (
                i
                for i, message in enumerate(user_messages)
                if message.id == record.last_user_message_id
            ),
            -1,
        )
        undelivered = (
            user_messages[delivered_index + 1 :]
            if delivered_index >= 0
            else user_messages[-1:]
        )
        for message in undelivered:
            text = _user_text(message).strip()
            if not text:
                continue
            events.append(
                {"type": "user.message", "content": [{"type": "text", "text": text}]}
            )
            last_user_message_id = message.id

        # The user moved on without answering the tools the frontend was asked
        # to run: fail those calls so the agent can respond to the new message.
        if last_user_message_id is not None and pending:
            abandoned = [
                {
                    "type": "user.custom_tool_result",
                    "custom_tool_use_id": tool_use_id,
                    "content": [
                        {
                            "type": "text",
                            "text": "The user did not provide a result for this tool call.",
                        }
                    ],
                    "is_error": True,
                }
                for tool_use_id in pending
            ]
            events[:0] = abandoned
            pending.clear()

        return _Outbound(
            events=events,
            still_parked=pending,
            last_user_message_id=last_user_message_id,
        )

    async def _get_or_create_session(
        self, thread_id: str, input: RunAgentInput, emit: Emit
    ) -> SessionRecord | None:  # noqa: A002
        existing = await maybe_await(self.store.get(thread_id))
        if existing is not None:
            return existing

        # A tool result only answers a pending call on an existing session;
        # never create a session to receive one.
        if not any(
            getattr(m, "role", None) == "user" and _user_text(m).strip()
            for m in input.messages
        ):
            return None

        # The busy-thread gate serializes runs per thread, so creation cannot race.
        record = await self._create_session(thread_id, input.tools or [])
        await maybe_await(self.store.set(thread_id, record))
        emit(
            CustomEvent(
                name="managed_agents.session",
                value={"sessionId": record.session_id, "threadId": thread_id},
            )
        )
        return record

    async def _create_session(
        self, thread_id: str, client_tools: Sequence[Any]
    ) -> SessionRecord:
        custom_tools = self._custom_tools(client_tools)
        title = (
            self.session_title(thread_id)
            if self.session_title
            else f"AG-UI thread {thread_id}"
        )

        agent: dict[str, Any]
        if not custom_tools:
            agent = {"type": "agent", "id": self.managed_agent_id}
            if self.agent_version is not None:
                agent["version"] = self.agent_version
        else:
            agent = {"type": "agent_with_overrides", "id": self.managed_agent_id}
            if self.agent_version is not None:
                agent["version"] = self.agent_version
            # Overrides replace the tool list, so keep the agent's own tools.
            agent["tools"] = [*(await self._base_tools()), *custom_tools]

        session = await self.client.beta.sessions.create(
            agent=agent, environment_id=self.environment_id, title=title
        )
        return SessionRecord(
            session_id=session.id, tool_names=[tool["name"] for tool in custom_tools]
        )

    def _custom_tools(self, client_tools: Sequence[Any]) -> list[dict[str, Any]]:
        """Frontend tools plus configured backend tools, as custom tool definitions.

        Keyed by normalized name. Distinct names that normalize alike never
        raise: the last one wins, and a frontend tool beats a backend tool with
        the same normalized name, matching dispatch order in the turn loop.
        """
        by_name: dict[str, dict[str, Any]] = {}
        for name, tool in self.backend_tools.items():
            by_name[name] = custom_tool_from(tool)
        for tool in client_tools:
            custom = custom_tool_from(tool)
            by_name[custom["name"]] = custom
        return list(by_name.values())

    async def _sync_client_tools(
        self, record: SessionRecord, client_tools: Sequence[Any]
    ) -> None:
        """Register any client tools the session's agent does not yet have.

        The tool list is a full replacement, so we merge with what the agent has.
        """
        desired = self._custom_tools(client_tools)
        known = set(record.tool_names)
        if all(tool["name"] in known for tool in desired):
            return
        await self.client.beta.sessions.update(
            record.session_id,
            agent={"tools": [*(await self._base_tools()), *desired]},
        )
        record.tool_names = [tool["name"] for tool in desired]

    async def _base_tools(self) -> list[Any]:
        """The tools defined on the managed agent itself, fetched fresh so console edits apply."""
        if self.agent_version is not None:
            agent = await self.client.beta.agents.retrieve(
                self.managed_agent_id, version=self.agent_version
            )
        else:
            agent = await self.client.beta.agents.retrieve(self.managed_agent_id)
        # The read shape is structurally compatible with the params shape.
        return list(getattr(agent, "tools", None) or [])

File: integrations/claude-managed-agents/typescript/src/agent.ts
import Anthropic from "@anthropic-ai/sdk";
import type { BetaManagedAgentsEventParams } from "@anthropic-ai/sdk/resources/beta/sessions/events";
import type { SessionCreateParams } from "@anthropic-ai/sdk/resources/beta/sessions/sessions";
import { AbstractAgent, EventType } from "@ag-ui/client";
import type { BaseEvent, Message, RunAgentInput, Tool } from "@ag-ui/client";
import { Observable } from "rxjs";
import { BEST_EFFORT_SEND_TIMEOUT_MS, DEFAULT_TURN_TIMEOUT_MS } from "./constants";
import { InMemorySessionStore } from "./sessions";
import { customToolFrom, customToolsFingerprint, normalizeToolName, type CustomToolParams } from "./tools";
import { runTurn, type TurnOutcome } from "./turn";
import type { BackendCustomTool, ManagedAgentsAgentConfig, SessionRecord, SessionStore } from "./types";

type OverrideTools = NonNullable<
  Extract<SessionCreateParams["agent"], { type: "agent_with_overrides" }>["tools"]
>;

/** The text of the given user message (string or multimodal content). */
const userText = (message: Extract<Message, { role: "user" }>): string => {
  if (typeof message.content === "string") return message.content;
  return (message.content ?? [])
    .map((part) => ("text" in part && typeof part.text === "string" ? part.text : ""))
    .join("");
};

/** Whether any user message in the run carries text. */
const hasUserText = (messages: Message[]): boolean =>
  messages.some((message) => message.role === "user" && userText(message).trim().length > 0);

/** A tool message's payload: its content plus any error text. */
const toolResultText = (message: Extract<Message, { role: "tool" }>): string =>
  [message.content, message.error].filter(Boolean).join("\n");

const runError = (message: string, code: string): BaseEvent =>
  ({ type: EventType.RUN_ERROR, message, code }) as BaseEvent;

const ABANDONED_TOOL_TEXT = "The user did not provide a result for this tool call.";

/**
 * An AG-UI agent backed by Claude Managed Agents. Each AG-UI thread maps to
 * one managed session; each run drives one turn of that session.
 */
export class ManagedAgentsAgent extends AbstractAgent {
  private readonly client: Anthropic;
  private readonly store: SessionStore;
  private readonly backendTools: Map<string, BackendCustomTool>;
  private currentRunAbort: AbortController | null = null;

  // Keyed by session-store identity: the store is the unit of tenancy, so
  // agents (and clones) sharing a store serialize runs per thread, while
  // per-caller stores keep one caller's runs from blocking another's.
  // Keys within a store's set are scoped to this managed agent.
  private static busyThreadsByStore = new WeakMap<SessionStore, Set<string>>();

  private get busyThreads(): Set<string> {
    let set = ManagedAgentsAgent.busyThreadsByStore.get(this.store);
    if (!set) {
      set = new Set<string>();
      ManagedAgentsAgent.busyThreadsByStore.set(this.store, set);
    }
    return set;
  }

  constructor(private config: ManagedAgentsAgentConfig) {
    super(config);
    this.client = config.client ?? new Anthropic();
    this.store = config.sessionStore ?? new InMemorySessionStore();
    this.backendTools = new Map((config.backendTools ?? []).map((tool) => [normalizeToolName(tool.name), tool]));
  }

  public clone() {
    // Clones share the client and session store so per-run copies see the
    // same thread↔session mappings.
    return new ManagedAgentsAgent({ ...this.config, client: this.client, sessionStore: this.store });
  }

  /** Report a swallowed failure. A broken hook must never break the run. */
  private report(operation: string, error: unknown, ids: { sessionId?: string; threadId?: string } = {}) {
    try {
      this.config.onError?.(error, { operation, ...ids });
    } catch {
      // ignored on purpose
    }
  }

  public abortRun() {
    this.currentRunAbort?.abort();
  }

  run(input: RunAgentInput): Observable<BaseEvent> {
    return new Observable<BaseEvent>((subscriber) => {
      const disconnect = new AbortController();
      this.currentRunAbort = disconnect;
      const timeoutMs = this.config.turnTimeoutMs ?? DEFAULT_TURN_TIMEOUT_MS;
      const timeout = AbortSignal.timeout(timeoutMs);
      const signal = AbortSignal.any([disconnect.signal, timeout]);
      const emit = (event: BaseEvent) => subscriber.next(event);

      this.runTurnForInput(input, emit, signal)
        .catch((err) => {
          if (disconnect.signal.aborted) return; // client went away; nothing left to tell
          if (timeout.aborted) {
            emit(runError(`The turn exceeded the ${timeoutMs / 1000}s limit and was interrupted.`, "turn_timeout"));
            return;
          }
          emit(runError(err instanceof Error && err.message ? err.message : "The run failed.", "run_failed"));
        })
        .finally(() => subscriber.complete());

      return () => {
        disconnect.abort();
        if (this.currentRunAbort === disconnect) this.currentRunAbort = null;
      };
    });
  }

  private async runTurnForInput(
    input: RunAgentInput,
    emit: (event: BaseEvent) => void,
    signal: AbortSignal,
  ): Promise<void> {
    // RunAgentInput is not validated at runtime; a body without `messages` or
    // `tools` must read as an empty run, not a TypeError.
    input = { ...input, messages: input.messages ?? [], tools: input.tools ?? [] };
    const { threadId, runId } = input;
    // One key for the session store and the busy-run gate, so a stored session
    // and the gate that serializes access to it can never disagree.
    const key = this.sessionKey(threadId);
    let sessionId: string | undefined;
    emit({ type: EventType.RUN_STARTED, threadId, runId } as BaseEvent);
    if (input.state !== undefined && input.state !== null) {
      emit({ type: EventType.STATE_SNAPSHOT, snapshot: input.state } as BaseEvent);
    }

    if (this.busyThreads.has(key)) {
      emit(runError("A run is already in progress on this thread.", "run_in_progress"));
      return;
    }
    // Check for something sendable before touching the API, so a malformed
    // run does not create an orphan session.
    if (!this.hasSendableContent(input.messages)) {
      emit(runError("There is nothing to send: this run has no user message or tool result.", "empty_run"));
      return;
    }
    this.busyThreads.add(key);
    try {
      const record = await this.getOrCreateSession(key, threadId, input, emit);
      if (!record) {
        emit(runError("There is nothing to send: a tool result arrived for a thread with no session.", "tool_result_without_session"));
        return;
      }
      sessionId = record.sessionId;
      await this.syncClientTools(record, input.tools);

      const outbound = this.outboundEvents(record, input.messages);
      if (outbound.events.length === 0) {
        emit(runError("There is nothing new to send: no user message or tool result in this run.", "nothing_to_send"));
        return;
      }

      // Some parked tool calls are still unanswered: post what we have and
      // stay parked instead of waiting on a session that will not resume.
      if (outbound.stillParked.length > 0) {
        await this.client.beta.sessions.events.send(record.sessionId, { events: outbound.events }, { signal });
        record.pendingClientToolUseIds = outbound.stillParked;
        record.lastUserMessageId = outbound.lastUserMessageId ?? record.lastUserMessageId;
        await this.store.set(key, record);
        emit({ type: EventType.RUN_FINISHED, threadId, runId } as BaseEvent);
        return;
      }

      const outcome = await runTurn({
        client: this.client,
        sessionId: record.sessionId,
        outbound: outbound.events,
        // Persist each delivery as soon as it lands, so a failure or
        // interruption later in the turn does not re-post it next run: the
        // tool results resume the session even if the follow-ups then fail.
        onResultsSent: async () => {
          record.pendingClientToolUseIds = [];
          await this.store.set(key, record);
        },
        onFollowUpsSent: async () => {
          if (outbound.lastUserMessageId) record.lastUserMessageId = outbound.lastUserMessageId;
          await this.store.set(key, record);
        },
        // Persist a park the moment the call is handed to the UI. A later
        // event can fail the turn before the session confirms the park, and
        // the remote session would then wait on an ID nothing remembers.
        onClientPark: async (toolUseId) => {
          if (record.pendingClientToolUseIds.includes(toolUseId)) return;
          record.pendingClientToolUseIds = [...record.pendingClientToolUseIds, toolUseId];
          await this.store.set(key, record);
        },
        clientTools: new Map((input.tools ?? []).map((tool) => [normalizeToolName(tool.name), tool.name])),
        backendTools: this.backendTools,
        toolConfirmation: this.config.toolConfirmation,
        streamDeltas: this.config.streamDeltas ?? true,
        onError: this.config.onError,
        emit,
        signal,
      });

      await this.recordOutcome(key, record, outcome);
      if (outcome.status !== "errored") {
        emit({ type: EventType.RUN_FINISHED, threadId, runId } as BaseEvent);
      }
    } catch (err) {
      // Interrupt the session while the busy gate is still held, so a user
      // who resends right away is not interrupted by this run's teardown.
      if (signal.aborted && sessionId) {
        await this.client.beta.sessions.events
          .send(sessionId, { events: [{ type: "user.interrupt" }] }, { signal: AbortSignal.timeout(BEST_EFFORT_SEND_TIMEOUT_MS) })
          .catch((error: unknown) => this.report("interrupt", error, { sessionId, threadId }));
      }
      throw err;
    } finally {
      this.busyThreads.delete(key);
    }
  }

  /** Whether the run carries a user message with text or a tool result. */
  private hasSendableContent(messages: Message[]): boolean {
    return hasUserText(messages) || messages.some((message) => message.role === "tool");
  }

  /**
   * The key that identifies this thread's state, in the session store and in
   * the busy-run gate alike, so two agents sharing one store neither adopt each
   * other's sessions nor serialize against each other's threads.
   *
   * Every field baked into the remote session at creation is part of the key:
   * none of them can be re-checked or changed on resume, so an agent must never
   * inherit a session created with a different environment, pinned version or
   * vault set. Each is length-prefixed so no two combinations can collide —
   * plain concatenation would let a `managedAgentId` of `support:beta` with
   * thread `t1` and one of `support` with thread `beta:t1` share one record.
   * The thread id is last, so it needs no prefix and may contain anything.
   */
  private sessionKey(threadId: string): string {
    const { managedAgentId, agentVersion, environmentId } = this.config;
    const field = (value: string) => `${value.length}:${value}|`;
    return (
      field(managedAgentId) +
      field(agentVersion === undefined ? "" : String(agentVersion)) +
      field(environmentId) +
      threadId
    );
  }

  /**
   * Reconcile the record with how the turn ended. An errored turn keeps
   * whatever `onClientPark` already persisted: the remote session is still
   * parked on those calls and the next run has to answer them.
   */
  private async recordOutcome(key: string, record: SessionRecord, outcome: TurnOutcome): Promise<void> {
    if (outcome.status === "errored") {
      if (outcome.sessionEnded) await this.store.delete(key);
      return;
    }
    if (outcome.status === "parked") {
      record.pendingClientToolUseIds = outcome.clientToolUseIds;
      await this.store.set(key, record);
      return;
    }
    // The session went idle on end_turn: nothing is awaited any more.
    if (record.pendingClientToolUseIds.length > 0) {
      record.pendingClientToolUseIds = [];
      await this.store.set(key, record);
    }
  }

  /**
   * Work out what to post into the session for this run: results for any
   * tool calls the frontend was asked to run, plus every user message not
   * yet delivered (in order).
   */
  private outboundEvents(
    record: SessionRecord,
    messages: Message[],
  ): { events: BetaManagedAgentsEventParams[]; stillParked: string[]; lastUserMessageId?: string } {
    const toolResult = (toolUseId: string, text: string, isError: boolean): BetaManagedAgentsEventParams => ({
      type: "user.custom_tool_result",
      custom_tool_use_id: toolUseId,
      content: [{ type: "text", text }],
      is_error: isError,
    });

    const answered: BetaManagedAgentsEventParams[] = [];
    const pending = new Set(record.pendingClientToolUseIds);
    for (const message of messages) {
      if (message.role !== "tool" || !pending.has(message.toolCallId)) continue;
      answered.push(toolResult(message.toolCallId, toolResultText(message), Boolean(message.error)));
      pending.delete(message.toolCallId);
    }

    // User messages after the last delivered one; on first contact, just the newest.
    let lastUserMessageId: string | undefined;
    const followUps: BetaManagedAgentsEventParams[] = [];
    const userMessages = messages.filter((message) => message.role === "user");
    const deliveredIndex = userMessages.findIndex((message) => message.id === record.lastUserMessageId);
    const undelivered = deliveredIndex >= 0 ? userMessages.slice(deliveredIndex + 1) : userMessages.slice(-1);
    for (const message of undelivered) {
      const text = userText(message).trim();
      if (!text) continue;
      followUps.push({ type: "user.message", content: [{ type: "text", text }] });
      lastUserMessageId = message.id;
    }

    // The user moved on without answering the tools the frontend was asked
    // to run: fail those calls (in their original order) so the agent can
    // respond to the new message.
    const abandoned: BetaManagedAgentsEventParams[] = [];
    if (lastUserMessageId !== undefined && pending.size > 0) {
      for (const toolUseId of pending) abandoned.push(toolResult(toolUseId, ABANDONED_TOOL_TEXT, true));
      pending.clear();
    }

    return { events: [...abandoned, ...answered, ...followUps], stillParked: [...pending], lastUserMessageId };
  }

  private async getOrCreateSession(
    key: string,
    threadId: string,
    input: RunAgentInput,
    emit: (event: BaseEvent) => void,
  ): Promise<SessionRecord | undefined> {
    const existing = await this.store.get(key);
    if (existing) return existing;

    // A tool result only answers a pending call on an existing session;
    // never create a session to receive one.
    if (!hasUserText(input.messages)) return undefined;

    // The busy-thread gate serializes runs per thread, so creation cannot race.
    const record = await this.createSession(threadId, input.tools ?? []);
    await this.store.set(key, record);
    emit({
      type: EventType.CUSTOM,
      name: "managed_agents.session",
      value: { sessionId: record.sessionId, threadId },
    } as BaseEvent);
    return record;
  }

  private async createSession(threadId: string, clientTools: Tool[]): Promise<SessionRecord> {
    const { managedAgentId, agentVersion, environmentId } = this.config;
    const customTools = this.customTools(clientTools);
    const title = this.config.sessionTitle?.(threadId) ?? `AG-UI thread ${threadId}`;
    const agentRef = { id: managedAgentId, ...(agentVersion !== undefined && { version: agentVersion }) };

    const agent: SessionCreateParams["agent"] =
      customTools.length === 0
        ? { type: "agent", ...agentRef }
        : { type: "agent_with_overrides", ...agentRef, tools: await this.mergedTools(customTools) };

    const session = await this.client.beta.sessions.create({ agent, environment_id: environmentId, title });
    return {
      sessionId: session.id,
      toolNames: customTools.map((tool) => tool.name),
      toolDefinitionsFingerprint: customToolsFingerprint(customTools),
      pendingClientToolUseIds: [],
    };
  }

  /**
   * Frontend tools plus configured backend tools, as custom tool definitions.
   * Keyed by normalized name: on a collision the last definition wins, and
   * a frontend tool always beats a backend tool of the same name, matching
   * dispatch order in the turn loop.
   */
  private customTools(clientTools: Tool[]): CustomToolParams[] {
    const byName = new Map<string, CustomToolParams>();
    for (const tool of this.config.backendTools ?? []) {
      byName.set(normalizeToolName(tool.name), customToolFrom(tool));
    }
    for (const tool of clientTools) {
      const custom = customToolFrom(tool);
      byName.set(custom.name, custom);
    }
    return [...byName.values()];
  }

  /** Keep the session's full replacement tool list aligned with this run. */
  private async syncClientTools(record: SessionRecord, clientTools: Tool[]): Promise<void> {
    const desired = this.customTools(clientTools);
    const fingerprint = customToolsFingerprint(desired);
    if (record.toolDefinitionsFingerprint === fingerprint) return;

    await this.client.beta.sessions.update(record.sessionId, {
      agent: { tools: await this.mergedTools(desired) },
    });
    record.toolNames = desired.map((tool) => tool.name);
    record.toolDefinitionsFingerprint = fingerprint;
  }

  /**
   * The agent's own tools plus `custom` tools, without duplicate names.
   * Overrides replace the whole list, so the agent's tools are carried
   * along, but a custom tool of the same name wins over the agent's copy.
   */
  private async mergedTools(custom: CustomToolParams[]): Promise<OverrideTools> {
    const names = new Set(custom.map((tool) => tool.name));
    const base = (await this.baseTools()).filter((tool) => tool.type !== "custom" || !names.has(tool.name));
    return [...base, ...custom];
  }

  /** The tools defined on the managed agent itself, fetched fresh so console edits apply. */
  private async baseTools(): Promise<OverrideTools> {
    const agent = await this.client.beta.agents.retrieve(
      this.config.managedAgentId,
      this.config.agentVersion !== undefined ? { version: this.config.agentVersion } : undefined,
    );
    // The read shape is structurally compatible with the params shape.
    return agent.tools as unknown as OverrideTools;
  }
}
